### DISCUSSION TITLE
Load and set `PORT` environment variable from `.env` file

### DISCUSSION THREAD

**Discussion by theoludwig** [vercel/next.js#29464] [Ideas]
### Describe the feature you'd like to request

There is often a `.env` file in Next.js projects, and when we use this file, we most likely ignore it in `.gitignore`, because it's quite specific to where and how we run our project.

We choose the port we want to run our application if the port is available on the current machine, so it is different for each computer and use case. To make it more convenient for everyone, that's why we often set `PORT` in `.env`.

So everyone is free to change it how they like.

### Describe the solution you'd like

I would like Next.js, load and set the `PORT` environment variable from the `.env` file automatically, by default (it could be a **BREAKING CHANGE**).

As stated in the [documentation](https://nextjs.org/docs/api-reference/cli) :
> Note: `PORT` can not be set in `.env` as booting up the HTTP server happens before any other code is initialized.

Make it happens that the code to load `env` is ran before booting up the HTTP server.

Related issues/PRs: https://github.com/vercel/next.js/pull/11408, https://github.com/vercel/next.js/issues/10338

### Describe alternatives you've considered

The current ways to change the default `PORT=3000` are :
- `-p, --port` CLI option: `npx next dev -p 4000`
- using the `PORT` environment variable: `PORT=4000 npx next dev`

The problem with these solutions is that we need to edit our `scripts` section of our `package.json` (we don't really need to, but it is way more convenient to start `npm run dev` than using `npx`).

Why is it problematic to update `package.json` to change the port?
Because we are often committing this file in `git` and fortunately we do, but the `PORT` is personal and not the same for everyone, to make it more convenient with everyone contributing to the project.

There is another way to do it, using the `PORT` environment variable as stated by this comment (https://github.com/vercel/next.js/pull/11408#issuecomment-617152924) :
```json
//package.json
{
  "scripts": {
    "start": "next start -p ${PORT-3000}"
  }
}
```

This has several drawbacks:

- it is not working on Windows, (again and again, if everyone could be on a Unix based system, the world would be better, but that's for another subject. :joy:)
I didn't test it myself, as I'm using Ubuntu GNU/Linux, but `${PORT-3000}`should not work on Windows.

- We need to export and load the PORT variable before running `npm run start`.

---

This issue/feature request is all about convenience than a really needed feature, as we can "workaround" quite easily, the idea is that Next.js should have good default, so we need less as possible to have config files and follow "standard" practices, there are several examples of frameworks doing this behavior as stated by this comment: https://github.com/vercel/next.js/issues/10338#issuecomment-672959041 and there are much more other examples.

**Comment by samuliasmala** [comment]
I wonder why the original issue https://github.com/vercel/next.js/issues/29454 was moved to discussions without any comments? The proposed feature would be very useful and also simple to implement. It is common to have deployment-specific configurations in `.env` file so having everything else in that file and port in environment variable makes configuration management unnecessarily complex.

**Comment by NandoMB** [comment]
Another way to do this "workaround" is using bash. It could be implemented at the start/dev script on NextCli...

*.env*
```env
HOSTNAME=localhost
PORT=4000
```

*package.json*
```json
"scripts": {
    "prestart": "chmod u+x ./server.development.sh && chmod u+x ./server.production.sh",
    "dev": "./server.development.sh",
    "build": "next build",
    "start": "./server.production.sh",
    "lint": "next lint"
},
```

*server.development.sh*
```sh
#!/bin/bash

if [ -f .env ]; then
  export $(echo $(cat .env | sed 's/#.*//g'| xargs) | envsubst)
fi

next dev --hostname $HOSTNAME --port $PORT
```

*server.production.sh*
```sh
#!/bin/bash

if [ -f .env ]; then
  export $(echo $(cat .env | sed 's/#.*//g'| xargs) | envsubst)
fi

next start --hostname $HOSTNAME --port $PORT
```

**Reply by samuliasmala** [reply]
FYI that export command doesn't work if  `.env` file contains entries with spaces. It's not an issue here since port or hostname won't have spaces but as a general `.env` file import I would use 

```bash
set -o allexport
source .env
set +o allexport
```

**Comment by SCG82** [comment]
Created a PR that adds an option to read various `.env` files on start/dev: https://github.com/vercel/next.js/pull/44392

**Comment by kachkaev** [comment]
I guess it’s a duplicate of https://github.com/vercel/next.js/issues/32603

**Comment by neilrackett** [comment]
Here's a solution based on a script we use to run Next.js 15 apps in Docker and Plesk environments, but with the addition of `dotenv`, that should work on all platforms:

`start.js`

```js
require('dotenv').config();

const cli = require('next/dist/cli/next-start');

cli.nextStart({
  port: process.env.PORT || 3000,
  hostname: process.env.HOSTNAME || '0.0.0.0',
});
```

`package.json`

```json
{
  "scripts": {
    "start": "node start.js"
  }
}
```

To use this with `dev`, just replace `next-start` and `nextStart` in the script with `next-dev` and `nextDev`.

**Reply by jdnielss** [reply]
this work for me btw, thanks

**Comment by sunstorymvp** [comment]
this works for me (linux, macos):

```json
"start": "export $(cat .env | xargs) && next dev --port $PORT --hostname $HOSTNAME",
```

**Comment by samislam** [comment]
Folks are no longer interested in this?

First download dotenv-cli from npm:
```bash
$ npm install  dotenv-cli --save-dev
```
Then create an environment variables file, call it whatever you like, for example I'm creating `.env.development` and `.env.production` to use them in different environments:
```bash
# make sure it's spelled PORT as bellow
PORT=4000 # put the port you want
```
Then simply change your npm script of `dev` command to the following:
```json
"scripts": {
  "dev": "dotenv -e .env.development -- next dev", # 👈 change the `-e` argument file path to your need.
  "build": "dotenv -e .env.production -- next build",
  "start": "dotenv -e .env.production -- next start",
}
```

Nextjs should take the environment variable automatically, I am running nextjs15 on Linux.
